src/main/scala/com/fulcrumgenomics/bam/TrimPrimers.scala

```diff
@@ -0,0 +1,195 @@
+
+            val (left, right) = if (r1.getReadNegativeStrandFlag) (r2, r1) else (r1, r2)
+            val (start, end ) = (left.getUnclippedStart, right.getUnclippedEnd)
+            val insert = new Interval(left.getContig, start, end)
+            val amp = detector.getOverlaps(insert).find(amp => abs(amp.leftStart - start) <= slop && abs(amp.rightEnd - end) <= slop)
+            amp match {
+              case Some(amplicon) =>
+                val leftClip  = amplicon.leftPrimerLength
+                val rightClip = amplicon.rightPrimerLength
+                reads.foreach { rec =>
+                  val toClip = if (rec.getFirstOfPairFlag == left.getFirstOfPairFlag) leftClip else rightClip
+                  SamRecordClipper.clip5PrimeEndOfRead(rec, toClip, mode)
+                }
+              case None =>
+                reads.foreach(r => SamRecordClipper.clip5PrimeEndOfRead(r, maxPrimerLength, mode))
+            }
+
+            clipFullyOverlappedFrReads(r1, r2)
+          }
+          // Pairs without both reads mapped in FR orientation are just maximally clipped
+          else {
+            reads.foreach(r => SamRecordClipper.clip5PrimeEndOfRead(r, maxPrimerLength, mode))
+          }
+
+          SamPairUtil.setMateInfo(r1, r2, true)
+          reads.filter(_.getSupplementaryAlignmentFlag).foreach { rec =>
+            val mate = if (rec.getFirstOfPairFlag) r2 else r1
+            SamPairUtil.setMateInformationOnSupplementalAlignment(rec, mate, true);
+          }
+        case _ =>
+          // Just trim each read independently
+          reads.foreach(r => SamRecordClipper.clip5PrimeEndOfRead(r, maxPrimerLength, mode))
+      }
+
+      reads.foreach(out.addAlignment)
+    }
+
+    out.close()
+  }
+
+  /** Gets an iterator in query name order over the records. */
+  private def queryNameOrderIterator(in: SamReader): BufferedIterator[SAMRecord] = {
+    if (in.getFileHeader.getSortOrder == SortOrder.queryname) {
+      in.iterator().bufferBetter
+    }
+    else {
+      logger.info("Sorting into queryname order.")
+      val sorter = SortingCollection.newInstance[SAMRecord](classOf[SAMRecord], new BAMRecordCodec(in.getFileHeader), new SAMRecordQueryNameComparator(), 1e6.toInt);
+      in.foreach(sorter.add)
+      sorter.iterator().bufferBetter
+    }
+  }
+
+  /** Fetches the next group of records that all share the same readname/template from the iterator. */
+  private def nextTemplate(iterator: BufferedIterator[SAMRecord]): Seq[SAMRecord] = {
+    val first    = iterator.next()
+    val template = first.getReadName
+    first :: iterator.takeWhile(_.getReadName == template).toList
+  }
+
+  /** Creates an overlap detector for all the amplicons from the input file. */
+  private def loadPrimerFile(path: FilePath): OverlapDetector[Amplicon] = {
+    val parser = DelimitedDataParser(path, '\t')
+    val detector = new OverlapDetector[Amplicon](0,0)
```

Comment:
I could, but I think it's probably clearer to keep it where the matching happens.